appdaemontestframework/hass_mocks.py

```diff
@@ -0,0 +1,130 @@
+            MockHandler(Hass, 'entity_exists'),
+        ]
+
+        # Generate a dictionary of mocked Hass functions for use by older code
+        # Note: This interface is considered deprecated and should be replaced with calls to public
+        # methods in the HassMocks object going forward.
+        self._hass_functions = {}
+        for mock_handler in self._mock_handlers:
+            self._hass_functions[mock_handler.function_name] = mock_handler.mock
+
+        # ensure compatibility with older versions of AppDaemon
+        self._hass_functions['passed_args'] = self._hass_functions['args']
+
+    ### Mock handling
+    def unpatch_mocks(self):
+        """Stops all mocks this class handles."""
+        for mock_handler in self._mock_handlers:
+            mock_handler.patch.stop()
+
+
+    ### Access to the deprecated hass_functions dict.
+    @property
+    def hass_functions(self):
+        return DeprecatedDict(self._hass_functions)
+
+    ### Logging mocks
+    @staticmethod
+    def _log_error(msg, level='ERROR'):
+        HassMocks._log_log(msg, level)
+
+    @staticmethod
+    def _log_log(msg, level='INFO'):
+        # Renamed the function to remove confusion
+        get_logging_level_from_name = logging.getLevelName
+        logging.log(get_logging_level_from_name(level), msg)
+
+
+class MockHandler:
+    """A class for generating a mock in an object and holding on to info about it.
+    :param object_to_patch: The object to patch
+    :param function_name: the name of the function to patch in the object
+    :param side_effect: side effect method to call. If not set, it will just return `None`
+    :param autospec: If `True` will autospec the Mock signature. Useful for getting `self` in side effects.
+    """
+    def __init__(self, object_to_patch, function_name, side_effect=None, autospec=False):
+        self.function_name = function_name
+        return_value = None
+        self.patch = mock.patch.object(object_to_patch, self.function_name, create=True,
+                                       autospec=autospec, side_effect=side_effect, return_value=None)
+        self.mock = self.patch.start()
+
+
+class WrappedMockHandler(MockHandler):
+    """Helper class that privides a 'wrapped' mock. This will automatically call the original function while still providing
+    a Mock for asserts and the such."""
+    def __init__(self, object_to_patch, function_name):
+        original_function = getattr(object_to_patch, function_name)
+        super().__init__(object_to_patch, function_name, side_effect=original_function, autospec=True)
+
+
+class DeprecatedDict(dict):
+    """Helper class that will give a deprectaion warning when accessing any of it's members"""
+    def __getitem__(self, key):
+        message = "Direct usage of `hass_functions['{}']` is deprecated in favor of direct use of the `HassMocks` object.".format(key)
+        warnings.warn(message, DeprecationWarning, stacklevel=2)
```

Comment:
As I was going through the review, I saw many times mention of _"Please now use **public methods of HassMocks**"_

But which public methods are you talking about exactly? 

Then I was thinking, instead of deprecating the complete usage of `hass_functions`, why not deprecate it the way it's accessed, but leave it available through `HassMocks`.  
That way there would be no need for any migration guide. Any call to `hass_functions` could be replaced by a call to `hass_mocks.hass_functions` (we'd need to remove the deprecation warning when accessing it this way).

But maybe `hass_functions` needed to be deprecated for a reason that will become clear in your next PRs. In that case, let's discuss about that migration guide I mentionned above 🙂

Reply:
My longer term thinking is that direct access to `hass_fucntions` shouldn't be needed in any tests in the future in favor of newer public methods in `hass_mocks`. The problem is, that this PR has yet to implement any of these new interfaces so there is no new suggestion yet.

Maybe this PR should be more about introducing `hass_mocks` and not yet introduce the deprecation of `hass_functions`? I might be missing some of the use cases for `hass_functions`. If I am (which I'm thinking is likely), replacing usage of `hass_functions` with `hass_mocks.hass_functions` might be the answer. Let me know what you think, but I'm going to stew on this for a bit and see if anything else comes to mind.

Reply:
Okay, I've stewed on this and come to the following thoughts:
1. `hass_mocks` is not a replacement for `hass_functions`
1. `hass_mocks` now contains `hass_functions`
1. As you suggest, allowing users to get at `hass_functions` through `hass_mocks` seems like an elegant way to handle this.
1. I think it still makes sense to deprecate the `hass_functions` test fixture in favor of `hass_mocks.hass_functions`. Access to the `hass_functions` test fixture can trigger a deprecation warning with explicit instructions on how to fix it.

This might take a bit more time than I have tonight to whip up, but let me know what you think if you read this before I get to try it. I'll post another comment when I push up a commit with some code to look at.

Reply:
Okay I took a stab at this. The new code will give a deprecation warning telling the user to switch over to the `hass_mocks` fixture and to now user `hass_mocks.hass_functions` to replace their hold code.

I think I was a little short sighted in the user cases of the end user to use the mocks directly in their tests. I still think a lot of those cases will be better serve by calls to `hass_mocks` public methods that do not yet exist. Since we wrap `hass_fuctions` as a property, this gives us allowance for fine grain control of deprecation messages in the future suggesting usage of these yet to be create functions (like time travel and the such).

Reply:
Amazing. You answered all my questions.
And YES, the deprecation message makes a lot of sense now! Thanks a lot 👍

Reply:
I wasn't super explicit what I meant is: I agree with your approach. It was confusing initially because I also couldn't see the "alternative methods in `HassMocks`" . . . but since they weren't there. . . 
Now it makes much more sense, and I agree that as functions are added in `HassMocks` we can gradually add deprecation instructions to the corresponding/replaced functions from `hass_functions`